### PR TITLE
feat(ui): support cron expression input

### PR DESCRIPTION
**Pull request by chirag-madlani**
JIRA Ticket: (TE-733)[https://cortexdata.atlassian.net/browse/TE-733]

Created cron-editor which supports quick schedule and custom cron expression

Screenshot:

Screenshot 2022-08-25 at 11 21 20 PM

Screenshot 2022-08-25 at 11 21 35 PM

### CODE REVIEW

**Comment by spham92** [comment]
For the UI cron builder we want to abstract away the actual cron string and make it more human readable.

Let's also have the option to paste a cron string like:

image

Where `Cron` is the original view to paste the cron string and `Day/Time` is the UI version

File: thirdeye-ui/src/app/components/alert-wizard-v2/alert-details/alert-frequency/alert-frequency.component.tsx [hunk 1]
```diff
@@ -46,10 +46,7 @@ function AlertFrequency<NewOrExistingAlert extends EditableAlert | Alert>({
                 </Box>
             </Grid>
 
-            <AlertDateTimeCronAdvance
-                cron={currentCron}
-                onCronChange={handleCronChange}
-            />
+            <CronEditor value={currentCron} onChange={handleCronChange} />
         </>
```
Comment (spham92) [inline thirdeye-ui/src/app/components/alert-wizard-v2/alert-details/alert-frequency/alert-frequency.component.tsx:50]: The switch between `Date/Time` and `Cron` should in this component and we should continue using `AlertDateTimeCronAdvance` for the `Cron` mode

File: thirdeye-ui/src/app/components/cron-editor/cron-editor.component.tsx [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+ * See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+import {
+    Box,
+    FormControl,
+    FormControlLabel,
+    FormHelperText,
+    Grid,
+    MenuItem,
```
Comment (spham92) [inline thirdeye-ui/src/app/components/cron-editor/cron-editor.component.tsx:20]: This component should solely be focused on converting the user friendly input into a cron string. It should also attempt to parse a passed cron string to see if it's simple enough to set the initial values for the hour, minute, and repeat every

File: thirdeye-ui/src/app/components/cron-editor/cron-editor.component.tsx [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+                                        </MenuItem>
+                                    ))}
+                                </TextField>
+                            </Grid>
+                        </Grid>
+                    </Grid>
+                    <CronHelper
+                        cron={dayTimeCron}
+                        isCronValid={isDayTimeCronValid}
+                    />
```
Comment (spham92) [inline thirdeye-ui/src/app/components/cron-editor/cron-editor.component.tsx:354]: This shouldn't be necessary. The user input fields should be intuitive enough

File: thirdeye-ui/src/app/components/cron-editor/cron-editor.component.tsx [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+                            <TextField
+                                fullWidth
+                                error={!isCronValid}
+                                value={cronInternal}
+                                variant="outlined"
+                                onChange={handleCronInputChange}
+                            />
+                        </Grid>
+                    </Grid>
+                    <CronHelper cron={cronInternal} isCronValid={isCronValid} />
```
Comment (spham92) [inline thirdeye-ui/src/app/components/cron-editor/cron-editor.component.tsx:370]: Remove this when this component is refactored to focus on getting user friendly input to convert to cron

File: thirdeye-ui/src/app/components/cron-helper/cron-helper.component.tsx [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+ * See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+import { FormHelperText, Grid, Link } from "@material-ui/core";
+import cronstrue from "cronstrue";
+import React from "react";
+import { useTranslation } from "react-i18next";
+import { CronHelperProps } from "./cron-helper.interface";
+
+const CronHelper: React.FunctionComponent<CronHelperProps> = ({
```
Comment (spham92) [inline thirdeye-ui/src/app/components/cron-helper/cron-helper.component.tsx:20]: This component should be unnecessary

**Comment by spham92** [comment]
For the single page forms, we should use https://github.com/startreedata/thirdeye/blob/master/thirdeye-ui/src/app/components/form-basics/input-section/input-section.component.tsx for the layout. See example usage here: https://github.com/startreedata/thirdeye/blob/master/thirdeye-ui/src/app/components/alert-wizard-v2/alert-details/alert-frequency/alert-date-time-cron-advance/alert-date-time-cron-advance.component.tsx

**Comment by spham92** [comment]
simplified the editor

image
image

**Comment by chirag-madlani** [comment]
@spham92 Are we looking for only days support here?

**Comment by spham92** [comment]
That's correct. We're just supporting the current feature set for now

**Review by spham92 — APPROVED** [review]

**Pull request merged**